cpp/src/arrow/dataset/scanner.h

```diff
@@ -39,27 +42,35 @@ struct ARROW_DS_EXPORT ScanContext {};
 class ARROW_DS_EXPORT ScanOptions {
  public:
   virtual ~ScanOptions() = default;
+
+  const DataSelector* selector() const { return selector_.get(); }
+
+ protected:
+  // Filters
+  std::unique_ptr<DataSelector> selector_;
 };
 
-/// \brief Read record batches from a range of a single data fragment
+/// \brief Read record batches from a range of a single data fragment. A
+/// ScanTask is meant to be a unit of work to be dispatched. The implementation
+/// must be thread and concurrent safe.
 class ARROW_DS_EXPORT ScanTask {
  public:
-  virtual ~ScanTask() = default;
-
   /// \brief Iterate through sequence of materialized record batches
   /// resulting from the Scan. Execution semantics encapsulated in the
   /// particular ScanTask implementation
   virtual std::unique_ptr<RecordBatchIterator> Scan() = 0;
+
+  virtual ~ScanTask() = default;
```

Comment:
Is there some reason to move the dtor to the bottom?

Reply:
I re-ordered by what I think is optimal for discovering (reading) the API.